### PR TITLE
[ALLUXIO-2232] Implementing File System REST API

### PR DESCRIPTION
**Pull request by jsimsa**
Addresses: https://alluxio.atlassian.net/browse/ALLUXIO-2232

In particular, this PR:

- introduces FileSystemClientRestServiceHandler, which implements file system REST API, deprecating the BlockMasterClientRestServiceHandler, BlockWorkerClientRestServiceHandler, FileSystemMasterClientRestServiceHandler, and LineageMasterClientRestServiceHandler
- introduces AlluxioProxy, a standalone Alluxio process which exposes the file system REST API
- adds GET_INFO endpoint for AlluxioMasterRestServiceHandler and AlluxioWorkerRestServiceHandler, which provides comprehensive information about Alluxio Master / Worker state, deprecating the individual endpoints these handlers provide
- performs minor clean up and renamings

### CODE REVIEW

**Comment by jsimsa** [comment]
Jenkins, test this please

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+      }
+    });
+  }
+
+  /**
+   * @summary upload a file
+   * @param path the file path
+   * @param blockSizeBytes the target block size in bytes
+   * @param mode the octal mode to used for file permissions
+   * @param persisted whether directory should be persisted
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: `file`? Should this be `WriteType` instead?

File: bin/alluxio-stop.sh [hunk 2]
```diff
@@ -19,17 +19,22 @@ BIN=$(cd "$( dirname "$0" )"; pwd)
 
 USAGE="Usage: alluxio-stop.sh [-h] [component]
 Where component is one of:
-  all\t\t\tStop local master/worker and remote workers. Default.
-  master\t\tStop local master.
-  worker\t\tStop local worker.
-  workers\t\tStop local worker and all remote workers.
+  all     \tStop local master/worker and remote workers.
```
Comment (jsimsa) [inline bin/alluxio-stop.sh]: done

File: core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java [hunk 3]
```diff
@@ -85,91 +88,100 @@
...
-   * @summary get the configuration map, the keys are ordered alphabetically.
+   * @summary get the Alluxio master information
    * @return the response object
    */
   @GET
-  @Path(GET_CONFIGURATION)
-  @ReturnType("java.util.SortedMap<java.lang.String, java.lang.String>")
-  public Response getConfiguration() {
-    return RestUtils.call(new RestUtils.RestCallable<Map<String, String>>() {
+  @Path(GET_INFO)
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java]: the problem with having separate endpoints is that the RPC overhead dominates ... I think a reasonable middle ground would be for the info to accept filter on what information should be filled in?

File: core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java [hunk 4]
```diff
@@ -85,91 +88,100 @@
...
+        return result;
       }
     });
   }
 
   /**
-   * @summary get the master rpc address
+   * @summary get the configuration map, the keys are ordered alphabetically.
    * @return the response object
+   * @deprecated since version 1.4 and will be removed in version 2.0
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java]: done

File: core/server/src/main/java/alluxio/proxy/AlluxioProxy.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+   * Blocks until the proxy is ready to serve requests.
+   */
+  public void waitForReady() {
+    while (true) {
+      if (mWebServer != null && mWebServer.getServer().isRunning()) {
+        return;
+      }
+      CommonUtils.sleepMs(10);
+    }
+  }
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/AlluxioProxy.java]: done and made the same change in AlluxioMaster

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 1 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: done

File: core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java (hunk 3 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java]: a filter sgtm

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 1 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: I also went through all the options and added ones that were missing (e.g. location policies)

File: core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java (hunk 3 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java]: I prefer to just call this GET_INFO since this is on the master REST service

File: core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java [hunk 6]
```diff
@@ -399,4 +435,60 @@ public Response getWorkerInfoList() {
...
+    for (Map.Entry<String, String> entry : properties) {
+      String key = entry.getKey();
+      if (PropertyKey.isValid(key)) {
+        configuration.put(key, entry.getValue());
+      }
+    }
+    return configuration;
+  }
+
+  private Map<String, Long> getMetricsInternal() {
```
Comment (uronce-cc) [inline core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java]: Manager frontend needs to distinguish between different types of metrics, e.g. RPC invocation counters, file system operations.

File: core/common/src/main/java/alluxio/Constants.java [hunk 7]
```diff
@@ -97,6 +99,7 @@
   public static final long LINEAGE_MASTER_WORKER_SERVICE_VERSION = 1;
   public static final long KEY_VALUE_MASTER_CLIENT_SERVICE_VERSION = 1;
   public static final long KEY_VALUE_WORKER_SERVICE_VERSION = 1;
+  public static final long FILE_SYSTEM_CLIENT_SERVICE_VERSION = 1;
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Constants.java]: we do not distinguish between the protocol type for other version constants ... for instance, BLOCK_MASTER_CLIENT_SERVICE_VERSION is used both in the Thrift block master client and REST block master client ... the service this version is for is "File System"

File: core/common/src/main/java/alluxio/Constants.java [hunk 8]
```diff
@@ -114,6 +117,7 @@
   public static final String FILE_SYSTEM_WORKER_CLIENT_SERVICE_NAME = "FileSystemWorkerClient";
   public static final String KEY_VALUE_MASTER_CLIENT_SERVICE_NAME = "KeyValueMasterClient";
   public static final String KEY_VALUE_WORKER_CLIENT_SERVICE_NAME = "KeyValueWorkerClient";
+  public static final String FILE_SYSTEM_CLIENT_SERVICE_NAME = "FileSystemClient";
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Constants.java]: ditto

File: core/common/src/main/java/alluxio/Constants.java (hunk 7 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/Constants.java]: ok, sgtm

File: core/common/src/main/java/alluxio/Constants.java (hunk 8 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/Constants.java]: sgtm

File: core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java (hunk 3 above)
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java]: Ah okay, sounds good.

File: core/common/src/main/java/alluxio/wire/AlluxioMasterInfo.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+package alluxio.wire;
+
+import com.google.common.base.Objects;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Alluxio master descriptor.
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/wire/AlluxioMasterInfo.java]: I prefer descriptor for consistency with Javadoc of other wire type (cf. alluxio.wire.FileInfo)

File: core/common/src/main/java/alluxio/wire/AlluxioMasterInfo.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+ */
+public class AlluxioMasterInfo {
+  private Capacity mCapacity;
+  private Map<String, String> mConfiguration;
+  private Map<String, Long> mMetrics;
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/wire/AlluxioMasterInfo.java]: there is currently no mechanism to do that ... the web UI hardcodes the selection ... we could add the categorization to the metrics system, but I would prefer to add do that in a separate PR

File: core/common/src/main/java/alluxio/wire/AlluxioWorkerInfo.java [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+package alluxio.wire;
+
+import com.google.common.base.Objects;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Alluxio worker descriptor.
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/wire/AlluxioWorkerInfo.java]: ditto

File: core/common/src/test/java/alluxio/wire/AlluxioMasterInfoTest.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+public class AlluxioMasterInfoTest {
+
```
Comment (jsimsa) [inline core/common/src/test/java/alluxio/wire/AlluxioMasterInfoTest.java]: done

File: core/common/src/test/java/alluxio/wire/CapacityTest.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+package alluxio.wire;
+
+import org.codehaus.jackson.map.ObjectMapper;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Random;
+
+public class CapacityTest {
+
```
Comment (jsimsa) [inline core/common/src/test/java/alluxio/wire/CapacityTest.java]: done

File: core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java [hunk 14]
```diff
@@ -53,8 +54,10 @@
 @Path(AlluxioMasterRestServiceHandler.SERVICE_PREFIX)
 @Produces(MediaType.APPLICATION_JSON)
 public final class AlluxioMasterRestServiceHandler {
-  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
   public static final String SERVICE_PREFIX = "master";
+
+  public static final String GET_INFO = "info";
+
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java]: I prefer to keep it to separate the deprecated endpoints from the GET_INFO endpoint

File: core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java (hunk 6 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java]: if we decide to do this, it should be added to the metrics system itself in a separate PR

File: core/server/src/main/java/alluxio/proxy/AlluxioProxy.java [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+  /**
+   * Creates an instance of {@link AlluxioProxy}.
+   */
+  public AlluxioProxy() {}
+
+  /**
+   * @return the actual bind hostname on web service (used by unit test only)
+   */
+  public String getWebBindHost() {
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/AlluxioProxy.java]: removed

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+   * @param path the file path
+   * @param loadMetadataType the {@link LoadMetadataType}. It overrides loadDirectChildren if it
+   *        is set.
+   * @return the response object
+   */
+  @GET
+  @Path(LIST_STATUS)
+  @ReturnType("java.util.List<alluxio.client.file.URIStatus>")
+  public Response listStatus(@QueryParam("path") final String path,
+      @DefaultValue("") @QueryParam("loadMetadataType") final String loadMetadataType) {
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: done

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+   * @param mode the file permission bits
+   * @param recursive whether the attribute should be set recursively
+   * @param ttlAction action to take after TTL is expired
+   * @return the response object
+   */
+  @POST
+  @Path(SET_ATTRIBUTE)
+  @ReturnType("java.lang.Void")
+  public Response setAttribute(@QueryParam("path") final String path,
+      @QueryParam("pinned") final Boolean pinned, @QueryParam("ttl") final Long ttl,
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: done the one-per-line ... as for alphabetization, I would prefer to keep the security options grouped

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+      @QueryParam("persisted") final Boolean persisted, @QueryParam("owner") final String owner,
+      @QueryParam("group") final String group, @QueryParam("mode") final Short mode,
+      @QueryParam("recursive") final Boolean recursive,
+      @QueryParam("ttxAction") final TtlAction ttlAction) {
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: good catch, thanks

File: core/server/src/main/java/alluxio/worker/AlluxioWorkerRestServiceHandler.java [hunk 19]
```diff
@@ -276,36 +298,95 @@ public Long call() throws Exception {
   /**
    * @summary get the worker metrics
    * @return the response object
+   * @deprecated since version 1.4 and will be removed in version 2.0
    */
   @GET
   @Path(GET_METRICS)
   @ReturnType("java.util.SortedMap<java.lang.String, java.lang.Long>")
+  @Deprecated
   public Response getMetrics() {
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/worker/AlluxioWorkerRestServiceHandler.java]: ditto

File: core/server/src/main/java/alluxio/worker/AlluxioWorkerRestServiceHandler.java [hunk 20]
```diff
@@ -276,36 +298,95 @@ public Long call() throws Exception {
...
+    for (Map.Entry<String, String> entry : properties) {
+      String key = entry.getKey();
+      if (PropertyKey.isValid(key)) {
+        configuration.put(key, entry.getValue());
+      }
+    }
+    return configuration;
+  }
+
+  private Map<String, Long> getMetricsInternal() {
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/worker/AlluxioWorkerRestServiceHandler.java]: ditto

File: tests/src/test/java/alluxio/rest/TestCaseOptions.java [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  private TestCaseOptions() {
+    mInputStream = null;
+    mJsonString = null;
+    mPrettyPrint = false;
+  }
+
+  /**
+   * @return the input stream
```
Comment (jsimsa) [inline tests/src/test/java/alluxio/rest/TestCaseOptions.java]: done

File: tests/src/test/java/alluxio/rest/TestCaseOptions.java [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+   */
+  public InputStream getInputStream() {
+    return mInputStream;
+  }
+
+  /**
+   * @return the JSON string
```
Comment (jsimsa) [inline tests/src/test/java/alluxio/rest/TestCaseOptions.java]: done

File: tests/src/test/java/alluxio/rest/TestCaseOptionsTest.java [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    options.setJsonString(jsonString);
+    options.setPrettyPrint(prettyPrint);
+
+    Assert.assertEquals(inputStream, options.getInputStream());
+    Assert.assertEquals(jsonString, options.getJsonString());
+    Assert.assertEquals(prettyPrint, options.isPrettyPrint());
+  }
+
+  @Test
+  public void equalsTest() throws Exception {
```
Comment (jsimsa) [inline tests/src/test/java/alluxio/rest/TestCaseOptionsTest.java]: done

File: core/common/src/main/java/alluxio/wire/AlluxioMasterInfo.java (hunk 9 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/wire/AlluxioMasterInfo.java]: Those javadoc are referring to class instances, not the class itself. "descriptor" implies something small like an object reference or primitive field, but this class has much more than that.

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 17 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: sgtm

File: core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java (hunk 14 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/master/AlluxioMasterRestServiceHandler.java]: I see, in that case maybe add a comment that the other endpoints are deprecated?

File: core/common/src/main/java/alluxio/wire/AlluxioMasterInfo.java (hunk 9 above)
Comment (jsimsa) [inline core/common/src/main/java/alluxio/wire/AlluxioMasterInfo.java]: changed all occurrences of `descriptor` under alluxio.wire to `information`

**Comment by jsimsa** [comment]
@aaudiber I agree with you that it would make sense to factor proxy out to a separate module, but I would prefer to do that in a separate PR

**Comment by jsimsa** [comment]
@aaudiber @calvinjia I addressed all your comments and fix the test crash issue, if this PR looks good to you, please merge it, thanks
